bindings/csharp/net4.6/Regorus.cs

```diff
@@ -1,27 +1,61 @@
+//-----------------------------------------------------------------------
+// <copyright file="Regorus.cs" company="Microsoft">
+//    Copyright (c)2012 Microsoft. All rights reserved.
+// </copyright>
+// <summary>
+//    Contains code for the Regorus Policy Engine base class.
+// </summary>
+//-----------------------------------------------------------------------
+
+
 using System;
 using System.Text;
+using System.IO;
 
-namespace Regorus
+namespace Microsoft.WindowsAzure.Regorus.IaaS
 {
-    public class Exception : System.Exception
-    {
-	public Exception(string message) : base(message) {}
-    }
-
-    public class Engine : ICloneable
-    {
+	//public class Exception : System.Exception
+	//{
+	//public Exception(string message) : base(message) {}
+	//}
+	public class PolicyEngine : ICloneable, IDisposable
+	{
 	unsafe private RegorusFFI.RegorusEngine* E;
-	public Engine()
+	
+	public PolicyEngine()
 	{
-	    unsafe
+		unsafe
 	    {
 		E = RegorusFFI.API.regorus_engine_new();
 	    }
 	}
 
+	private bool disposed = false;
+
+	public void Dispose()
+	{
+		Dispose(true);
+		GC.SuppressFinalize(this); // Prevent the destructor from being called
+	}
+	
+	protected virtual void Dispose(bool disposing)
+	{
+		if (!disposed)
+		{
+			if (disposing)
```

Comment:
I happened to see this PR by chance and have some questions about multi-threading:
- Can `Dispose` be called simultaneously from two threads?
- Can the same PolicyEngine instance be used from multiple threads?

Reply:
Dispose could be called by two threads but they would belong to different class object. In other words, each policy engine would be used in a single thread only.

Reply:
> Dispose could be called by two threads but they would belong to different class object

Is this guaranteed by C#?

> The introduction of the dispose() is to ensure the timely cleanup of the engine to avoid complexity in a multi-thread environment.

I don't quite understand the need for `disposed`. It seems to indicate that Dispose can be called on the same object more than once.

On a related note, in a language like C++ if the code is written as
```
if (!disposed) {
   // do something
   disposed = true;
}
```

the compiler is free to reorder and rewrite it as
```
if (!disposed) {
   disposed = true;
   // do something
}
```
unless special care is taken to prevent the compiler from doing so. Not sure how it behaves in C#.

Reply:
there is no way to prevent dispose() to be called multiple times. the code is to ensure it is only called once because I am not sure if there are any side effects to call RegorusFFI.API.regorus_engine_drop(E) multiple times. will it have any side effects? I will modify the code you pointed out as the current code doesn't look right to me now. :)

Reply:
> > Dispose could be called by two threads but they would belong to different class object
> 
> Is this guaranteed by C#?
yes, if you want to have determined time to clean up an object before its destructed, dispose() can be used. 

> > The introduction of the dispose() is to ensure the timely cleanup of the engine to avoid complexity in a multi-thread environment.
> 
> I don't quite understand the need for `disposed`. It seems to indicate that Dispose can be called on the same object more than once.
there is no way to prevent dispose() to be called multiple times.  It is our job to make sure if that happens, no side effect is there. 

> On a related note, in a language like C++ if the code is written as
> 
> ```
> if (!disposed) {
>    // do something
>    disposed = true;
> }
> ```
> 
> the compiler is free to reorder and rewrite it as
> 
> ```
> if (!disposed) {
>    disposed = true;
>    // do something
> }
> ```
> 
> unless special care is taken to prevent the compiler from doing so. Not sure how it behaves in C#.

I will modify the code you pointed out as the current code doesn't look right to me now. :)